### PR TITLE
[DRAFT] [M7] New backend to propagate OCM requests to Hive

### PR DESCRIPTION
**Pull request by mbarnes**
**Note:** This PR builds upon https://github.com/Azure/ARO-RP/pull/2332 and is therefore in draft state ~until the first PR is merged~.
**Update 9/14:** https://github.com/Azure/ARO-RP/pull/2332 is merged but leaving this in **DRAFT** until [SD-ADR-0077](https://docs.google.com/document/d/1CU4Cmjz3nh643bSAlPEfrfpJeAa_xMBVmg4lDQo1n90/) is decided.

### Which issue this PR addresses:

Implements feature: [M7 - ARO Hive Day-2 SyncSet Mechanism](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/14846838)

### What this PR does / why we need it:

Implements a new backend and a new CosmosDB container to store backend state.  One and only one RP instance (per region) processes events from a CosmosDB change feed and propagates the changes to the region's Hive cluster on AKS.

### Test plan for issue:

To meet the required deadline, only manual testing was possible.

### Is there any documentation that needs to be updated for this PR?

No, but some rationale for the backend design is captured in commit messages.

### CODE REVIEW

File: pkg/backend/clustermanagerconfiguration.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	docs, err := cmc.configIterator.Next(ctx, -1)
+	if err != nil {
+		return false, err
+	}
+
+	workWasDone := false
+
+	if docs != nil {
```
Comment (petrkotas) [inline pkg/backend/clustermanagerconfiguration.go:77]: In this loop I see it takes the configuration and passes it directly to the AKS/Hive cluster. What is the plan for linking the configuration with RP clusterDocument? I can easily see AKS going bananas and the need to restore the state. We also need a way to know which configuration is linked to which cluster.

Pardon me, if I missed it, this PR is realy big.

**Review by petrkotas — CHANGES_REQUESTED** [review]
Hi @mbarnes thank you for your work. It is really a lot of effort put into this PR. I have a question related to the document handling bellow.

In general, how is the overal document flow? It is hard to understand the relation between backendDocument and clusterManagerConfiguration.
Would you have a diagram with such info? Thank you a lot and thank you again for the work put into this PR.

File: pkg/database/backends.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	collClient := cosmosdb.NewCollectionClient(dbClient, dbid)
+	triggerClient := cosmosdb.NewTriggerClient(collClient, collBackends)
+
+	trigger := &cosmosdb.Trigger{
+		ID:               renewLeaseTriggerID,
+		TriggerOperation: cosmosdb.TriggerOperationAll,
+		TriggerType:      cosmosdb.TriggerTypePre,
+		Body: fmt.Sprintf(`function trigger() {
```
[the remark this answers is not in the document]
Comment (mbarnes) [reply pkg/database/backends.go:52]: It is, I stole that from `pkg/database/monitors.go`.

File: pkg/backend/clustermanagerconfiguration.go (hunk 1 above)
Comment (mbarnes) [reply pkg/backend/clustermanagerconfiguration.go:77]: Yeah, it's subtle.  The assumption is the payload from OCM will include the correct Hive namespace for the target cluster.  All we have to do is forward it on.

So if we had to do disaster recovery, I expect we would first recreate the same Hive namespaces for existing clusters based on the RP `clusterdocument`s.  Then we just throw all the resources in the `ClusterManagerConfigurations` container at Hive and they would land in the correct namespaces.

**Comment by mbarnes** [comment]
> It is hard to understand the relation between backendDocument and clusterManagerConfiguration.

The two documents are mostly unrelated.  `backendDocument` is just a place for backends to stash state.  Backend state in this case means 1) a lease that only one RP instance can acquire, and 2) the so-called "continuation value" for the `clusterManagerConfiguration` change feed, which is basically a bookmark.

So if an RP instance holding the `backendDocument` lease dies for whatever reason, the lease will expire after one minute and another RP instance can then acquire the lease and begin reading the `clusterManagerConfiguration` change feed starting from the last recorded "continuation value" (otherwise it would replay the entire change history of the DB container).

Does that help clarify?

**Comment by mociarain** [comment]
> #2332 is merged but leaving this in **DRAFT** until [SD-ADR-0077](https://docs.google.com/document/d/1CU4Cmjz3nh643bSAlPEfrfpJeAa_xMBVmg4lDQo1n90/) is decided.

Looks like that ADR has been accepted. What do we want to do with this guy now?

**Comment by mbarnes** [comment]
Sorry, I didn't realize this was still open.  Closing since all this code is obsolete.

As I understand, at this point OCM integration with ARO Classic is unlikely to ever happen.  The API extensions for Hive resources from OCM (#2332) should never have been published because we're stuck with this unused public API now.

**Pull request closed without merging**